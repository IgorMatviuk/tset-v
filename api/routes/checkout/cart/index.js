const cart = require('../../../fixtures/cart')

function checkout(req, res) {
  res.jsonp({
    data: cart
  })
}

module.exports = checkout
