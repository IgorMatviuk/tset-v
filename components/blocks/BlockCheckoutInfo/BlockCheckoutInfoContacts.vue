<template>
  <div class="checkout-info">
    <div class="checkout-info__title">{{ $t('checkoutPage.contactInfo') }}</div>
    <!-- <BFormGroup class="checkout-info__group">
      <BFormRadioGroup
        v-model="value.contactInfo.isGuest.$model"
        class="checkout-info__radio-group"
        name="status"
      >
        <BFormRadio
          :class="{ error: value.contactInfo.isGuest.$error }"
          class="checkout-info__radio"
          value="true"
        >
          Новый покупатель
        </BFormRadio>
        <BFormRadio
          :class="{ error: value.contactInfo.isGuest.$error }"
          class="checkout-info__radio"
          value="false"
        >
          Постоянный клиент
        </BFormRadio>
      </BFormRadioGroup>
    </BFormGroup> -->
    <BFormGroup
      class="checkout-info__group custom-field-group"
      :label="$t('form.labels.first_name')"
      label-for="checkoutFirstName"
      label-class="checkout-info__label custom-label custom-label--required"
      label-cols-md="3"
    >
      <BFormInput
        id="checkoutFirstName"
        v-model.trim="value.contactInfo.firstName.$model"
        :class="{ error: value.contactInfo.firstName.$error }"
        class="checkout-info__input"
        type="text"
        :trim="true"
        :placeholder="$t('form.placeholders.first_name')"
      />
      <template v-if="value.contactInfo.firstName.$error">
        <div
          v-if="getShowModelError('firstName', 'required')"
          class="error-message"
        >
          {{ $t('form.errors.first_name.required') }}
        </div>
        <div
          v-else-if="getShowModelError('firstName', 'minLength')"
          class="error-message"
        >
          {{ $t('form.errors.first_name.minLength') }}
        </div>
      </template>
    </BFormGroup>
    <BFormGroup
      class="checkout-info__group custom-field-group"
      :label="$t('form.labels.last_name')"
      label-for="checkoutLastName"
      label-class="checkout-info__label custom-label custom-label--required"
      label-cols-md="3"
    >
      <BFormInput
        id="checkoutLastName"
        v-model.trim="value.contactInfo.lastName.$model"
        :class="{ error: value.contactInfo.lastName.$error }"
        class="checkout-info__input"
        type="text"
        :trim="true"
        :placeholder="$t('form.placeholders.last_name')"
      />
      <template v-if="value.contactInfo.lastName.$error">
        <div
          v-if="getShowModelError('lastName', 'required')"
          class="error-message"
        >
          {{ $t('form.errors.last_name.required') }}
        </div>
        <div
          v-else-if="getShowModelError('lastName', 'minLength')"
          class="error-message"
        >
          {{ $t('form.errors.last_name.minLength') }}
        </div>
      </template>
    </BFormGroup>
    <BFormGroup
      class="checkout-info__group custom-field-group"
      :label="$t('form.labels.phone')"
      label-for="checkoutPhone"
      label-class="checkout-info__label custom-label custom-label--required"
      label-cols-md="3"
    >
      <client-only>
        <vue-tel-input
          ref="tel"
          v-model.trim="value.contactInfo.phone.$model"
          v-bind="phoneOptions"
          :wrapper-classes="
            value.contactInfo.phone.$error
              ? 'form-control-phone form-control error'
              : 'form-control-phone form-control'
          "
          @country-changed="onPhoneCountryChange($event)"
        />
      </client-only>
      <template v-if="value.contactInfo.phone.$error">
        <div
          v-if="getShowModelError('phone', 'required')"
          class="error-message"
        >
          {{ $t('form.errors.phone.required') }}
        </div>
        <div
          v-else-if="getShowModelError('phone', 'phone')"
          class="error-message"
        >
          {{ $t('form.errors.phone.phone') }}
        </div>
      </template>
    </BFormGroup>
    <BFormGroup
      class="checkout-info__group custom-field-group"
      :label="$t('form.labels.email')"
      label-for="checkoutEmail"
      label-class="checkout-info__label custom-label custom-label--required"
      label-cols-md="3"
    >
      <BFormInput
        id="checkoutEmail"
        v-model.trim="value.contactInfo.email.$model"
        :class="{ error: value.contactInfo.email.$error }"
        class="checkout-info__input"
        type="email"
        :trim="true"
        :placeholder="$t('form.placeholders.email')"
      />
      <template v-if="value.contactInfo.email.$error">
        <div
          v-if="getShowModelError('email', 'required')"
          class="error-message"
        >
          {{ $t('form.errors.email.required') }}
        </div>
        <div
          v-else-if="getShowModelError('email', 'email')"
          class="error-message"
        >
          {{ $t('form.errors.email.email') }}
        </div>
      </template>
    </BFormGroup>
    <BFormGroup
      class="checkout-info__group custom-field-group"
      :label="$t('form.labels.comment')"
      label-for="checkoutComment"
      label-class="checkout-info__label custom-label"
      label-cols="12"
      label-cols-md="3"
    >
      <BFormTextarea
        id="checkoutComment"
        v-model.trim="value.contactInfo.comment.$model"
        :class="{ error: value.contactInfo.comment.$error }"
        class="checkout-info__input"
        type="text"
        :trim="true"
        :placeholder="$t('form.placeholders.comment')"
      />
    </BFormGroup>
  </div>
</template>

<script>
// import phoneCountry from '@/mixins/phone-country'

export default {
  // mixins: [phoneCountry],
  props: {
    value: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      phoneOptions: {
        inputId: 'checkoutPhone',
        inputClasses: 'checkout-info__input form-control',
        placeholder: this.$t('form.placeholders.phone'),
        validCharactersOnly: true,
        defaultCountry: 'UA',
        mode: 'international'
      }
    }
  },
  methods: {
    getShowModelError(fieldname, validatorName) {
      const validator = this.value.contactInfo[fieldname][validatorName]
      return typeof validator !== 'undefined' && !validator
    },
    onPhoneCountryChange(activeCountry) {
      /* -------- DO NOT REMOVE THIS! HARDCODE! -------- */
      /* istanbul ignore else */
      if (activeCountry.iso2 !== 'UA') {
        this.$emit('non-defeault-country')
      }
    }
  }
}
</script>

<style lang="scss">
.checkout-info {
  &__title {
    margin-bottom: 20px;
    font-size: 14px;
    line-height: 18px;
    text-transform: uppercase;
    letter-spacing: 0.05em;
  }

  &__radio {
    &:not(:last-of-type) {
      margin-right: 25px;
    }
  }

  &__label {
    flex: 0 0 24%;
    max-width: 24%;
  }

  &__input {
    font-size: 13px;
    line-height: 16px;
    letter-spacing: 0.05em;
  }

  @include media-breakpoint-down(xl) {
    &__label {
      flex: 0 0 29%;
      max-width: 29%;
    }
  }

  @include media-breakpoint-down(lg) {
    &__label {
      flex: 0 0 32%;
      max-width: 32%;
    }

    &__radio {
      &:not(:last-of-type) {
        margin-right: 26px;
      }
    }
  }

  @include media-breakpoint-down(md) {
    &__radio {
      &:not(:last-of-type) {
        margin-right: 22px;
      }
    }

    &__label {
      flex: 0 0 17.1%;
      max-width: 17.1%;
    }
  }

  @include media-breakpoint-down(sm) {
    &__title {
      margin-bottom: 14px;
    }

    &__label {
      flex: 0 0 100%;
      max-width: 100%;
    }

    &__group {
      margin-bottom: 12px;
    }
  }
}
</style>
