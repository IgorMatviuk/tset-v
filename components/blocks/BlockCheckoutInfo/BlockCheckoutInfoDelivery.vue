<template>
  <div class="checkout-info">
    <div class="checkout-info__title">
      {{ $t('checkoutPage.deliveryMethod') }}
    </div>
    <BFormGroup class="checkout-info__group checkout-info__delivery-methods">
      <BFormRadioGroup
        v-model="value.deliveryInfo.deliveryMethod.$model"
        class="checkout-info__radio-group"
        name="delivery_method"
      >
        <BFormRadio
          :class="{ error: value.deliveryInfo.deliveryMethod.$error }"
          class="checkout-info__radio checkout-info__radio--delivery"
          value="novaposhta"
        >
          {{ $t('checkoutPage.delivery.novaposhta') }}
        </BFormRadio>
        <BFormRadio
          :class="{ error: value.deliveryInfo.deliveryMethod.$error }"
          class="checkout-info__radio checkout-info__radio--delivery"
          value="local"
        >
          {{ $t('checkoutPage.delivery.local') }}
        </BFormRadio>
        <BFormRadio
          :class="{ error: value.deliveryInfo.deliveryMethod.$error }"
          class="checkout-info__radio checkout-info__radio--delivery"
          value="abroad"
        >
          {{ $t('checkoutPage.delivery.abroad') }}
        </BFormRadio>
      </BFormRadioGroup>
    </BFormGroup>
    <div class="checkout-info__delivery-wrapper">
      <BFormGroup
        v-if="abroadDelivery"
        class="checkout-info__group custom-field-group"
        :label="$t('form.labels.country')"
        label-for="checkoutСountry"
        label-class="checkout-info__label checkout-info__label--delivery custom-label custom-label--required"
        label-cols-md="3"
      >
        <SelectWithSearch
          v-model="value.deliveryInfo.deliveryCountry"
          :options="countryOptions"
          field-id="checkoutСountry"
          option-field="name"
          assign-field="id"
          :placeholder="$t('form.placeholders.country')"
        />
      </BFormGroup>
      <template v-if="!abroadDelivery">
        <BFormGroup
          class="checkout-info__group custom-field-group"
          :label="$t('form.labels.region')"
          label-for="checkoutRegion"
          label-cols-md="3"
          :label-class="cityLabelClass"
        >
          <SelectWithSearch
            v-model="value.deliveryInfo.deliveryRegion"
            :options="regionOptions"
            field-id="checkoutRegion"
            option-field="name"
            assign-field="id"
            :placeholder="$t('form.placeholders.region')"
            @selected="onRegionSelected"
          />
        </BFormGroup>
        <BFormGroup
          :class="{ 'checkout-info__group--disabled': noRegionSelected }"
          class="checkout-info__group custom-field-group"
          :label="$t('form.labels.state')"
          label-for="checkoutState"
          label-cols-md="3"
          :label-class="cityLabelClass"
        >
          <SelectWithSearch
            v-model="value.deliveryInfo.deliveryState"
            :options="stateOptions"
            field-id="checkoutState"
            option-field="name"
            assign-field="id"
            :placeholder="$t('form.placeholders.state')"
            @selected="onStateSelected"
          />
        </BFormGroup>
      </template>
      <BFormGroup
        v-else
        class="checkout-info__group custom-field-group"
        :label="$t('form.labels.state')"
        label-for="checkoutСity"
        label-cols-md="3"
        :label-class="cityLabelClass"
      >
        <BFormInput
          id="checkoutCity"
          v-model.trim="value.deliveryInfo.deliveryCity.$model"
          :class="{ error: value.deliveryInfo.deliveryCity.$error }"
          class="checkout-info__input"
          type="text"
          :trim="true"
          :placeholder="$t('form.placeholders.state')"
        />
      </BFormGroup>
      <BFormGroup
        v-if="postalDelivery"
        :class="{ 'checkout-info__group--disabled': noStateSelected }"
        class="checkout-info__group custom-field-group"
        :label="$t('form.labels.office')"
        label-for="checkoutOffice"
        label-class="checkout-info__label checkout-info__label--delivery custom-label custom-label--required"
        label-cols-md="3"
      >
        <SelectWithSearch
          v-model="value.deliveryInfo.deliveryOffice"
          :options="officeOptions"
          field-id="checkoutOffice"
          option-field="name"
          assign-field="id"
          :placeholder="$t('form.placeholders.office')"
        />
      </BFormGroup>
      <BFormGroup
        v-if="targetDelivery"
        :class="{
          'checkout-info__group--disabled': noStateSelected
        }"
        class="checkout-info__group custom-field-group"
        :label="$t('form.labels.address')"
        label-for="checkoutAddress"
        :label-class="addressLabelClass"
        label-cols-md="3"
      >
        <BFormInput
          id="checkoutAddress"
          v-model.trim="value.deliveryInfo.deliveryAddress.$model"
          :class="{ error: value.deliveryInfo.deliveryAddress.$error }"
          class="checkout-info__input"
          type="text"
          :trim="true"
          :placeholder="$t('form.placeholders.address')"
        />
        <div
          v-if="value.deliveryInfo.deliveryAddress.$error"
          class="error-message"
        >
          {{ $t('form.errors.address.required') }}
        </div>
      </BFormGroup>
    </div>
    <div class="checkout-info__delivery-sale">
      {{ $t('checkoutPage.freeShipping.message') }}
    </div>
  </div>
</template>

<script>
import { mapState } from 'vuex'

import SelectWithSearch from '@/components/shared/SelectWithSearch'

import { actionTypes as cartActionTypes } from '@/store/cart'

import { DELIVERY_METHODS } from '@/utils/constants'

export default {
  components: {
    SelectWithSearch
  },
  props: {
    value: {
      type: Object,
      required: true
    }
  },
  computed: {
    ...mapState({
      countryOptions: (state) => state.cart.countries,
      regionOptions: (state) => state.cart.regions,
      stateOptions: (state) => state.cart.states,
      officeOptions: (state) => state.cart.offices
    }),
    abroadDelivery() {
      return (
        this.value.deliveryInfo.deliveryMethod.$model ===
        DELIVERY_METHODS.ABROAD
      )
    },
    postalDelivery() {
      return (
        this.value.deliveryInfo.deliveryMethod.$model ===
        DELIVERY_METHODS.NOVAPOSHTA
      )
    },
    targetDelivery() {
      return (
        this.value.deliveryInfo.deliveryMethod.$model ===
          DELIVERY_METHODS.LOCAL ||
        this.value.deliveryInfo.deliveryMethod.$model ===
          DELIVERY_METHODS.ABROAD
      )
    },
    cityLabelClass() {
      return this.value.deliveryInfo.deliveryMethod.$model ===
        DELIVERY_METHODS.NOVAPOSHTA ||
        this.value.deliveryInfo.deliveryMethod.$model === DELIVERY_METHODS.LOCAL
        ? 'checkout-info__label checkout-info__label--delivery custom-label custom-label--required'
        : 'checkout-info__label checkout-info__label--delivery custom-label'
    },
    addressLabelClass() {
      return this.value.deliveryInfo.deliveryMethod.$model ===
        DELIVERY_METHODS.LOCAL
        ? 'checkout-info__label checkout-info__label--delivery custom-label custom-label--required'
        : 'checkout-info__label checkout-info__label--delivery custom-label'
    },
    noRegionSelected() {
      return (
        this.value.deliveryInfo.deliveryRegion &&
        !this.value.deliveryInfo.deliveryRegion.$model &&
        this.value.deliveryInfo.deliveryMethod.$model !==
          DELIVERY_METHODS.ABROAD
      )
    },
    noStateSelected() {
      return (
        this.value.deliveryInfo.deliveryState &&
        !this.value.deliveryInfo.deliveryState.$model &&
        this.value.deliveryInfo.deliveryMethod.$model !==
          DELIVERY_METHODS.ABROAD
      )
    }
  },
  methods: {
    async onRegionSelected(region) {
      this.$emit('region-change')

      const deliveryMethod = this.value.deliveryInfo.deliveryMethod.$model

      /* istanbul ignore else */
      if (deliveryMethod !== DELIVERY_METHODS.ABROAD) {
        await this.$store.dispatch(
          `cart/${cartActionTypes.GET_STATES}`,
          region.id
        )
      }
    },
    async onStateSelected(state) {
      this.$emit('state-change')

      const deliveryMethod = this.value.deliveryInfo.deliveryMethod.$model

      /* istanbul ignore else */
      if (deliveryMethod === DELIVERY_METHODS.NOVAPOSHTA) {
        await this.$store.dispatch(
          `cart/${cartActionTypes.GET_OFFICES}`,
          state.id
        )
      }
    }
  }
}
</script>

<style lang="scss">
.checkout-info {
  &__radio {
    &--delivery {
      margin-bottom: 10px;
    }
  }

  &__group--disabled {
    pointer-events: none;
    cursor: not-allowed;
    opacity: 0.5;
  }

  &__delivery-wrapper {
    margin-top: -10px;
  }

  &__delivery-sale {
    margin-top: 28px;
    font-size: 12px;
    line-height: 15px;
    color: $primary;
  }

  @include media-breakpoint-down(md) {
    &__label {
      &--delivery {
        flex: 0 0 25.6%;
        max-width: 25.6%;
      }
    }
  }

  @include media-breakpoint-down(sm) {
    &__label {
      &--delivery {
        flex: 0 0 100%;
        max-width: 100%;
      }
    }

    &__delivery-sale {
      margin-top: 16px;
      font-size: 11px;
      line-height: 14px;
    }
  }
}
</style>
