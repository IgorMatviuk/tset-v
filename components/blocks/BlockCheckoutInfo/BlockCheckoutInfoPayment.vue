<template>
  <div class="checkout-info">
    <div class="checkout-info__title">
      {{ $t('checkoutPage.paymentMethod') }}
    </div>
    <BFormGroup class="checkout-info__group">
      <BFormRadioGroup
        v-model="value.paymentInfo.paymentMethod.$model"
        class="checkout-info__radio-group"
        name="payment-method"
      >
        <BFormRadio
          v-if="isAbroadDelivery"
          :class="{ error: value.paymentInfo.paymentMethod.$error }"
          class="checkout-info__radio checkout-info__radio--payment"
          value="cod"
        >
          {{ $t('checkoutPage.paymentMethods.cod') }}
        </BFormRadio>
        <BFormRadio
          v-if="isAbroadDelivery"
          :class="{ error: value.paymentInfo.paymentMethod.$error }"
          class="checkout-info__radio checkout-info__radio--payment"
          value="privat24"
        >
          {{ $t('checkoutPage.paymentMethods.privat') }}
        </BFormRadio>
        <BFormRadio
          :class="{ error: value.paymentInfo.paymentMethod.$error }"
          class="checkout-info__radio checkout-info__radio--payment"
          value="liqPay"
        >
          Credit Card / Debit Card (LiqPay)
        </BFormRadio>
      </BFormRadioGroup>
    </BFormGroup>
  </div>
</template>

<script>
export default {
  props: {
    value: {
      type: Object,
      required: true
    },
    isAbroadDelivery: {
      type: Boolean,
      required: true
    }
  }
}
</script>

<style lang="scss">
.checkout-info {
  &__radio {
    &--payment {
      width: 100%;

      &:not(:last-of-type) {
        margin-right: 0;
        margin-bottom: 12px;
      }
    }
  }

  @include media-breakpoint-down(sm) {
    &__radio {
      &--payment {
        &:not(:last-of-type) {
          margin-bottom: 8px;
        }
      }
    }
  }
}
</style>
