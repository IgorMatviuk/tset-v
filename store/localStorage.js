export const mutationTypes = {
  SET_CART_ID: 'SET_CART_ID'
}

export const state = () => ({
  cartId: null,
  status: false
})

export const mutations = {
  [mutationTypes.SET_CART_ID](state, cartId) {
    state.cartId = cartId
  }
}
