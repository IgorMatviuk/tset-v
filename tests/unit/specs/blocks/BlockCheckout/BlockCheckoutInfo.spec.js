import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import { BButton, BFormGroup } from 'bootstrap-vue'
import _cloneDeep from 'lodash.clonedeep'

import BlockCheckoutInfo from '@/components/blocks/BlockCheckoutInfo'
import BlockCheckoutInfoContacts from '@/components/blocks/BlockCheckoutInfo/BlockCheckoutInfoContacts'
import BlockCheckoutInfoDelivery from '@/components/blocks/BlockCheckoutInfo/BlockCheckoutInfoDelivery'
import BlockCheckoutInfoPayment from '@/components/blocks/BlockCheckoutInfo/BlockCheckoutInfoPayment'

import { state as cartState, mutations as cartMutations } from '@/store/cart'
import { mutationTypes as modalMutationTypes } from '@/store/modal'
import { mutationTypes as localStorageMutationTypes } from '@/store/localStorage'

import { DELIVERY_METHODS, PAYMENT_METHODS } from '@/utils/constants'

import regionsFixture from '@/api/fixtures/regions'
import statesFixture from '@/api/fixtures/states'
import officesFixture from '@/api/fixtures/offices'

import '@/tests/unit/mocks/html-form-element'

const localVue = createLocalVue()
localVue.component('BButton', BButton)
localVue.component('BFormGroup', BFormGroup)
localVue.use(Vuex)

describe('BlockCheckoutInfo', () => {
  let store
  let mockAxios
  let mockAxiosFallback
  let mockPostRequest
  let mockRouterPush
  let mockI18nPath
  let mockSetCartId
  let mockSetModalMeta
  let mockSetModalVisibilityToggle

  function createCheckoutInfo() {
    const wrapper = mount(BlockCheckoutInfo, {
      localVue,
      store,
      propsData: {
        isVisible: true,
        checkoutStep: 1,
        checkoutSteps: {
          orderSummary: 1,
          contactInfo: 2,
          deliveryInfo: 3,
          paymentInfo: 4
        }
      },
      mocks: {
        $t: (str) => str,
        $axios: mockAxios || mockAxiosFallback,
        $router: {
          push: mockRouterPush
        },
        $i18n: {
          path: mockI18nPath
        },
        $v: {
          $invalid: false,
          $touch: () => {},
          $reset: () => {},
          formData: {}
        }
      },
      stubs: [
        'svg-icon',
        'client-only',
        'NuxtLink',
        'BlockCheckoutInfoContacts',
        'BlockCheckoutInfoDelivery',
        'BlockCheckoutInfoPayment'
      ]
    })

    return wrapper
  }

  function setDeliveryMethod(wrapper, method = DELIVERY_METHODS.NOVAPOSHTA) {
    wrapper.setData({
      formData: { deliveryInfo: { deliveryMethod: method } }
    })
  }

  function setDeliveryCountry(wrapper, country = 'Country_Mock') {
    wrapper.setData({
      formData: { deliveryInfo: { deliveryCountry: country } }
    })
  }

  function setDeliveryAddress(wrapper, address = 'Address_Mock') {
    wrapper.setData({
      formData: { deliveryInfo: { deliveryAddress: address } }
    })
  }

  function setDeliveryRegion(wrapper, region = 'Region_Mock') {
    wrapper.setData({
      formData: { deliveryInfo: { deliveryRegion: region } }
    })
  }

  function setDeliveryState(wrapper, state = 'State_Mock') {
    wrapper.setData({
      formData: { deliveryInfo: { deliveryState: state } }
    })
  }

  function setDeliveryOffice(wrapper, office = 'Office_Mock') {
    setDeliveryRegion(wrapper)
    setDeliveryState(wrapper)

    wrapper.setData({
      formData: {
        deliveryInfo: {
          deliveryOffice: office
        }
      }
    })
  }

  function setPaymentMethod(wrapper, method) {
    wrapper.setData({
      formData: {
        paymentInfo: {
          paymentMethod: method
        }
      }
    })
  }

  beforeAll(() => {
    mockSetCartId = jest.fn()
    mockSetModalMeta = jest.fn()
    mockSetModalVisibilityToggle = jest.fn()

    store = new Vuex.Store({
      modules: {
        localStorage: {
          namespaced: true,
          state: {
            cartId: null
          },
          mutations: {
            [localStorageMutationTypes.SET_CART_ID]: mockSetCartId
          }
        },
        cart: {
          namespaced: true,
          state: {
            ...cartState(),
            regions: _cloneDeep(regionsFixture),
            states: _cloneDeep(statesFixture),
            offices: _cloneDeep(officesFixture)
          },
          mutations: { ...cartMutations }
        },
        modal: {
          namespaced: true,
          mutations: {
            [modalMutationTypes.SET_MODAL_META]: mockSetModalMeta,
            [modalMutationTypes.SET_MODAL_VISIBILITY_TOGGLE]: mockSetModalVisibilityToggle
          }
        }
      }
    })

    mockAxiosFallback = {
      $post: mockPostRequest
    }

    mockPostRequest = jest.fn()

    mockRouterPush = jest.fn()

    mockI18nPath = jest.fn(() => '/')
  })

  describe('Structure', () => {
    test('has expected structure', () => {
      const wrapper = createCheckoutInfo()

      const contacts = wrapper.findComponent(BlockCheckoutInfoContacts)
      expect(contacts.exists()).toBe(true)
      const delivery = wrapper.findComponent(BlockCheckoutInfoDelivery)
      expect(delivery.exists()).toBe(true)
      const payment = wrapper.findComponent(BlockCheckoutInfoPayment)
      expect(payment.exists()).toBe(true)
    })
  })

  describe('Delivery info', () => {
    test('country change', () => {
      const wrapper = createCheckoutInfo()
      const { deliveryInfo } = wrapper.vm.formData

      setDeliveryCountry(wrapper)

      expect(deliveryInfo.deliveryCountry).toBe('Country_Mock')

      wrapper.vm.onCountryChange()

      expect(deliveryInfo.deliveryCity).toBeNull()
    })

    test('region change', () => {
      const wrapper = createCheckoutInfo()
      const { deliveryInfo } = wrapper.vm.formData

      setDeliveryState(wrapper)
      setDeliveryOffice(wrapper)

      expect(deliveryInfo.deliveryState).toBe('State_Mock')
      expect(deliveryInfo.deliveryOffice).toBe('Office_Mock')

      wrapper.vm.onRegionChange()

      expect(deliveryInfo.deliveryState).toBeNull()
      expect(deliveryInfo.deliveryOffice).toBeNull()
    })

    test('state change', () => {
      const wrapper = createCheckoutInfo()
      const { deliveryInfo } = wrapper.vm.formData

      setDeliveryOffice(wrapper)

      expect(deliveryInfo.deliveryOffice).toBe('Office_Mock')

      wrapper.vm.onStateChange()

      expect(deliveryInfo.deliveryOffice).toBeNull()
    })

    test('abroad delivery only', () => {
      const wrapper = createCheckoutInfo()
      const { deliveryInfo } = wrapper.vm.formData

      setDeliveryMethod(wrapper, DELIVERY_METHODS.NOVAPOSHTA)

      expect(deliveryInfo.deliveryMethod).toBe(DELIVERY_METHODS.NOVAPOSHTA)

      wrapper.vm.setAbroadDeliveryOnly()

      expect(deliveryInfo.deliveryMethod).toBe(DELIVERY_METHODS.ABROAD)
    })
  })

  describe('Validation rules', () => {
    test('using "novaposhta" delivery method', () => {
      // NOTE: "novaposhta" is a default value
      const wrapper = createCheckoutInfo()
      const validations = wrapper.vm.$options.validations.bind(wrapper.vm)

      expect(validations()).toMatchSnapshot()
    })

    test('using "local" delivery method', () => {
      const wrapper = createCheckoutInfo()
      const validations = wrapper.vm.$options.validations.bind(wrapper.vm)

      setDeliveryMethod(wrapper, DELIVERY_METHODS.LOCAL)
      expect(validations()).toMatchSnapshot()
    })

    test('using "abroad" delivery method', () => {
      const wrapper = createCheckoutInfo()
      const validations = wrapper.vm.$options.validations.bind(wrapper.vm)

      setDeliveryMethod(wrapper, DELIVERY_METHODS.ABROAD)
      expect(validations()).toMatchSnapshot()
    })
  })

  describe('Form Submition', () => {
    beforeEach(() => {
      mockAxios = {
        $post: jest.fn().mockResolvedValue({ data: { order_id: 1 } })
      }
    })

    async function placeOrder(wrapper, { deliveryMethod, paymentMethod }) {
      fillFormData(wrapper, deliveryMethod)
      await wrapper.vm.$nextTick()

      switch (deliveryMethod) {
        case DELIVERY_METHODS.NOVAPOSHTA:
          setDeliveryOffice(wrapper)
          break
        case DELIVERY_METHODS.LOCAL:
          setDeliveryRegion(wrapper)
          setDeliveryState(wrapper)
          setDeliveryAddress(wrapper)
          break
        case DELIVERY_METHODS.ABROAD:
          setDeliveryCountry(wrapper)
          break
        default:
          throw new Error('Invalid delivery method.')
      }

      setPaymentMethod(wrapper, paymentMethod)
      await wrapper.vm.$nextTick()

      wrapper.find('form').trigger('submit')
    }

    function fillFormData(wrapper, deliveryMethod) {
      // TODO: click on "new user" radio button
      // wrapper.find("[value='new-user']").trigger('click')

      wrapper.setData({
        formData: {
          contactInfo: {
            firstName: 'John',
            lastName: 'Doe',
            phone: '+380 66 123 4567',
            email: 'user@example.com'
          },
          deliveryInfo: {
            deliveryMethod
          }
        }
      })
    }

    describe('Using "novaposhta" delivery method', () => {
      test('using "COD" payment method', async () => {
        const wrapper = createCheckoutInfo()
        await placeOrder(wrapper, {
          deliveryMethod: DELIVERY_METHODS.NOVAPOSHTA,
          paymentMethod: PAYMENT_METHODS.COD
        })
      })

      test('using "Privat24" payment method', async () => {
        const wrapper = createCheckoutInfo()
        await placeOrder(wrapper, {
          deliveryMethod: DELIVERY_METHODS.NOVAPOSHTA,
          paymentMethod: PAYMENT_METHODS.PRIVAT
        })
      })

      test('using "LiqPay" payment method', async () => {
        const wrapper = createCheckoutInfo()
        await placeOrder(wrapper, {
          deliveryMethod: DELIVERY_METHODS.NOVAPOSHTA,
          paymentMethod: PAYMENT_METHODS.LIQPAY
        })
      })
    })

    describe('Using "local" delivery method', () => {
      test('using "COD" payment method', async () => {
        const wrapper = createCheckoutInfo()
        await placeOrder(wrapper, {
          deliveryMethod: DELIVERY_METHODS.LOCAL,
          paymentMethod: PAYMENT_METHODS.COD
        })
      })

      test('using "Privat24" payment method', async () => {
        const wrapper = createCheckoutInfo()
        await placeOrder(wrapper, {
          deliveryMethod: DELIVERY_METHODS.LOCAL,
          paymentMethod: PAYMENT_METHODS.PRIVAT
        })
      })

      test('using "LiqPay" payment method', async () => {
        const wrapper = createCheckoutInfo()
        await placeOrder(wrapper, {
          deliveryMethod: DELIVERY_METHODS.LOCAL,
          paymentMethod: PAYMENT_METHODS.LIQPAY
        })
      })
    })

    describe('Using "abroad" delivery method', () => {
      test('using "LiqPay" payment method', async () => {
        const wrapper = createCheckoutInfo()
        await placeOrder(wrapper, {
          deliveryMethod: DELIVERY_METHODS.ABROAD,
          paymentMethod: PAYMENT_METHODS.LIQPAY
        })
      })
    })
  })

  describe('Mobile navigation', () => {
    test('emit "next-step" event', () => {
      const wrapper = createCheckoutInfo()

      wrapper.vm.goToNextStep()

      expect(wrapper.emitted()['next-step'].length).toBe(1)
    })
  })
})
