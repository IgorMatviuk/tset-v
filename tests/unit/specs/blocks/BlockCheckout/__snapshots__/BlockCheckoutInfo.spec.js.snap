// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlockCheckoutInfo Validation rules using "abroad" delivery method 1`] = `
Object {
  "formData": Object {
    "contactInfo": Object {
      "comment": Object {
        "onlyText": [Function],
      },
      "email": Object {
        "email": [Function],
        "required": [Function],
      },
      "firstName": Object {
        "minLength": [Function],
        "onlyLetters": [Function],
        "required": [Function],
      },
      "lastName": Object {
        "minLength": [Function],
        "onlyLetters": [Function],
        "required": [Function],
      },
      "phone": Object {
        "phone": [Function],
        "required": [Function],
      },
    },
    "deliveryInfo": Object {
      "deliveryAddress": Object {},
      "deliveryCity": Object {},
      "deliveryCountry": Object {
        "required": [Function],
      },
      "deliveryMethod": Object {
        "required": [Function],
      },
      "deliveryOffice": Object {},
    },
    "paymentInfo": Object {
      "paymentMethod": Object {
        "required": [Function],
      },
    },
  },
}
`;

exports[`BlockCheckoutInfo Validation rules using "local" delivery method 1`] = `
Object {
  "formData": Object {
    "contactInfo": Object {
      "comment": Object {
        "onlyText": [Function],
      },
      "email": Object {
        "email": [Function],
        "required": [Function],
      },
      "firstName": Object {
        "minLength": [Function],
        "onlyLetters": [Function],
        "required": [Function],
      },
      "lastName": Object {
        "minLength": [Function],
        "onlyLetters": [Function],
        "required": [Function],
      },
      "phone": Object {
        "phone": [Function],
        "required": [Function],
      },
    },
    "deliveryInfo": Object {
      "deliveryAddress": Object {
        "required": [Function],
      },
      "deliveryCountry": Object {},
      "deliveryMethod": Object {
        "required": [Function],
      },
      "deliveryOffice": Object {},
      "deliveryRegion": Object {
        "required": [Function],
      },
      "deliveryState": Object {
        "required": [Function],
      },
    },
    "paymentInfo": Object {
      "paymentMethod": Object {
        "required": [Function],
      },
    },
  },
}
`;

exports[`BlockCheckoutInfo Validation rules using "novaposhta" delivery method 1`] = `
Object {
  "formData": Object {
    "contactInfo": Object {
      "comment": Object {
        "onlyText": [Function],
      },
      "email": Object {
        "email": [Function],
        "required": [Function],
      },
      "firstName": Object {
        "minLength": [Function],
        "onlyLetters": [Function],
        "required": [Function],
      },
      "lastName": Object {
        "minLength": [Function],
        "onlyLetters": [Function],
        "required": [Function],
      },
      "phone": Object {
        "phone": [Function],
        "required": [Function],
      },
    },
    "deliveryInfo": Object {
      "deliveryAddress": Object {},
      "deliveryCountry": Object {},
      "deliveryMethod": Object {
        "required": [Function],
      },
      "deliveryOffice": Object {
        "required": [Function],
      },
      "deliveryRegion": Object {
        "required": [Function],
      },
      "deliveryState": Object {
        "required": [Function],
      },
    },
    "paymentInfo": Object {
      "paymentMethod": Object {
        "required": [Function],
      },
    },
  },
}
`;
